src/test/libraryfilesynctest.cpp

```diff
@@ -0,0 +1,712 @@
+        ASSERT_TRUE(file.open(QIODevice::ReadWrite | QIODevice::ExistingOnly | QIODevice::Append));
+        const auto fileSize = file.size();
+        // Grow and shrink the file to bump the modification time stamp
+        ASSERT_TRUE(file.resize(fileSize + 1));
+        ASSERT_TRUE(file.flush());
+        ASSERT_TRUE(file.resize(fileSize));
+        file.close();
+        const auto newLastModified = fileLastModified();
+        ASSERT_TRUE(newLastModified.isValid());
+        ASSERT_GT(newLastModified, oldLastModified);
+        sleepAfterFileLastModifiedUpdated();
+    }
+
+    void removeFile() const {
+        ASSERT_TRUE(m_fileInfo.toQFile().remove());
+    }
+
+  private:
+    void createFile(const QFileInfo& testFile) {
+        ASSERT_TRUE(m_tempDir.isValid());
+        ASSERT_FALSE(m_fileInfo.exists());
+        mixxxtest::copyFile(
+                testFile.absoluteFilePath(),
+                m_fileInfo.location());
+        ASSERT_TRUE(m_fileInfo.exists());
+        sleepAfterFileLastModifiedUpdated();
+    }
+
+    const QTemporaryDir m_tempDir;
+    const mixxx::FileInfo m_fileInfo;
+};
+
+class SyncTrackMetadataConfigScope final {
+  public:
+    explicit SyncTrackMetadataConfigScope(
+            UserSettingsPointer pConfig)
+            : m_pConfig(std::move(pConfig)),
+              m_syncTrackMetadataBackup(
+                      m_pConfig->getValue(
+                              mixxx::library::prefs::kSyncTrackMetadataConfigKey,
+                              false)) {
+    }
+    ~SyncTrackMetadataConfigScope() {
+        // Restore config
+        m_pConfig->setValue(
+                mixxx::library::prefs::kSyncTrackMetadataConfigKey,
+                m_syncTrackMetadataBackup);
+    }
+
+    void setConfig(bool syncTrackMetadata) const {
+        m_pConfig->setValue(
+                mixxx::library::prefs::kSyncTrackMetadataConfigKey,
+                syncTrackMetadata);
+    }
+
+  private:
+    const UserSettingsPointer m_pConfig;
+    const bool m_syncTrackMetadataBackup;
+};
+
+} // namespace
+
+class LibraryFileSyncTest : public LibraryTest {
+  protected:
+    explicit LibraryFileSyncTest(const QFileInfo& testFile)
+            : m_tempFileSystem(testFile) {
+        createTrack();
+    }
+
+    TrackPointer loadTrack() const {
+        const auto pTrack = trackCollectionManager()->getTrackById(m_trackId);
+        DEBUG_ASSERT(pTrack);
+        return pTrack;
+    }
+
+    void saveModifiedTrack(TrackPointer&& pTrack) const {
+        ASSERT_NE(nullptr, pTrack);
+        ASSERT_EQ(1, pTrack.use_count());
+        const auto trackId = pTrack->getId();
+        ASSERT_TRUE(trackId.isValid());
+        ASSERT_TRUE(pTrack->isDirty());
+        pTrack.reset();
+        ASSERT_EQ(nullptr, GlobalTrackCacheLocker().lookupTrackById(trackId));
+        // The file might have been modified if track metadata has been exported.
+        sleepAfterFileLastModifiedUpdated();
```

Comment:
Can this be removed? It looks like it is wrong here.

Reply:
I don't think so. We have to insert a sleep whenever file system time stamps are modified to ensure... Please read my code comments. I am not going to repeat them here.

Reply:
I don't understand the comment here. Can you improve it? 

It is used in two cases, an in both cases it does not shift following times. it just delay the lookup of the just written file time. 

If this has an effect something else must go on. Maybe is a aching issue or such.

Reply:
The sleep only affects any following write operation. Reading the modified time stamp does not need to be delayed, but the sleep delays everything.

The comment says exactly what happens, i.e. a write operation to the file system might happen behind the scenes depending on the circumstances.

Reply:
Caching issues have already been fixed and file times are always read directly in the test.